routing/bicycle_directions.cpp

```diff
@@ -222,39 +213,172 @@ Index::FeaturesLoaderGuard & BicycleDirectionsEngine::GetLoader(MwmSet::MwmId co
+      // In this case that turn candidate angle is invalid and
+      // should not be used for turn generation.
+      adjacentEdges.m_outgoingTurns.isCandidatesAngleValid = false;
+    }
+    adjacentEdges.m_outgoingTurns.candidates.emplace_back(angle, uniNodeId, highwayClass);
+  }
+}
+
+Segment BicycleDirectionsEngine::GetSegment(FeatureID const & featureId, uint32_t segId, bool forward) const
+{
+  CHECK(m_numMwmIds, ());
+
+  if (!featureId.m_mwmId.IsAlive())
+    MYTHROW(RoutingException, ("Mwm:", featureId.m_mwmId, "is not alive."));
+
+  NumMwmId const numMwmId =
+    m_numMwmIds->GetId(featureId.m_mwmId.GetInfo()->GetLocalFile().GetCountryFile());
+  return Segment(numMwmId, featureId.m_index, segId, forward);
+}
+
+void BicycleDirectionsEngine::FillPathSegmentsAndAdjacentEdgesMap(
+    RoadGraphBase const & graph, vector<Junction> const & path,
+    IRoadGraph::TEdgeVector const & routeEdges, my::Cancellable const & cancellable)
+{
+  size_t const pathSize = path.size();
+  CHECK_GREATER(pathSize, 1, ());
+  CHECK_EQUAL(routeEdges.size(), pathSize - 1, ());
+  // Filling |m_adjacentEdges|.
+  m_adjacentEdges.insert(make_pair(UniNodeId(UniNodeId::Type::Mwm), AdjacentEdges(0)));
+  auto constexpr kInvalidSegId = std::numeric_limits<uint32_t>::max();
+  // |startSegId| is a value to keep start segment id of a new instance of LoadedPathSegment.
+  uint32_t startSegId = kInvalidSegId;
+  vector<Junction> prevJunctions;
+  vector<Segment> prevSegments;
+  for (size_t i = 1; i < pathSize; ++i)
+  {
+    if (cancellable.IsCancelled())
+      return;
+
+    Junction const & prevJunction = path[i - 1];
+    Junction const & currJunction = path[i];
+
+    // Note. i + 1 == pathSize means |currJunction| is a finish and the outgoing edges
+    // from finish are not important for turn generation.
+    IRoadGraph::TEdgeVector outgoingEdges;
+    if (i + 1 != pathSize)
+      graph.GetOutgoingEdges(currJunction, outgoingEdges);
+
+    IRoadGraph::TEdgeVector ingoingEdges;
+    graph.GetIngoingEdges(currJunction, ingoingEdges);
+
+    Edge const & inEdge = routeEdges[i - 1];
+    // Note. |inFeatureId| may be invalid in case of adding fake features.
+    // It happens for example near starts and a finishes.
+    FeatureID const & inFeatureId = inEdge.GetFeatureId();
+    uint32_t const inSegId = inEdge.GetSegId();
+    bool const inIsForward = inEdge.IsForward();
+
+    if (startSegId == kInvalidSegId)
+      startSegId = inSegId;
+
+    if (inFeatureId.IsValid() && i + 1 != pathSize &&
+        !IsJoint(ingoingEdges, outgoingEdges, inEdge, routeEdges[i]))
+    {
+      prevJunctions.push_back(prevJunction);
+      if (m_generateTrafficSegs && !inEdge.IsFake())
+        prevSegments.push_back(GetSegment(inFeatureId, inSegId, inIsForward));
+      continue;
+    }
+    CHECK_EQUAL(prevJunctions.size(), abs(static_cast<int32_t>(inSegId - startSegId)), ());
+
+    prevJunctions.push_back(prevJunction);
```

Comment:
Yes. It's done. But now I'm trying to rewrite this loop to two loop according to your former comment.